### ISSUE TITLE
Different Graph generations

### ISSUE THREAD

**Issue by amithrm** [pytorch/xla#4994]
## 🐛 Bug

This code snippet is extracted from the AdamW optimizer. This optimizer for different ranges of learning rate and weight decay generates different graphs. This is causing unexpected compilations during the running of the application. The fix is also mentioned in the section. However, such scenarios can occur anywhere and we need a generic mechanism to make sure the same graph is generated. 

## To Reproduce
```
import torch
import torch, random, os
import numpy as np
import torch_xla.core.xla_model as xm

os.environ["NEURON_FRAMEWORK_DEBUG"] = "1"
os.environ["XLA_IR_DEBUG"] = "1"
os.environ["XLA_HLO_DEBUG"]="1"
os.environ['XLA_USE_BF16']="1"
os.environ['XLA_NO_SPECIAL_SCALARS']="1"

def func1():
    param = torch.FloatTensor([0.001]).to(xm.xla_device())
    lr = 2.9999999999999997e-06
    weight_decay = 0.01
    param.mul_(1 - lr * weight_decay)
    print(param)

def func2():
    param = torch.FloatTensor([0.001]).to(xm.xla_device())
    lr = 4.6874999999999995e-08
    weight_decay = 0.01
    param.mul_(1 - lr * weight_decay)
    print(param)

def func3():
    param = torch.FloatTensor([0.001]).to(xm.xla_device())
    lr = 2.9999999999999997e-06
    weight_decay = 0.01
    param.sub_(param * lr * weight_decay)
    print(param)

def func4():
    param1 = torch.FloatTensor([0.001]).to(xm.xla_device())
    lr1 = 4.6874999999999995e-08
    weight_decay1 = 0.01
    param1.sub_(param1 * lr1 * weight_decay1)
    print(param1)

func1()
func2()
func3()
func4()
```

<!--
It is really important for the team to have a quick repro, which requires no setup work.

The quicker is the repro to be run, the higher the chances the bug will be addressed sooner.

The best way to create quick repros is to create a Colab based on the following template:

```
https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information

Things to avoid in repros is the need to download datasets which require setting up keys or other login information, like Kaggle downloads for example.

Another example are Colab which mount user's Google Drive storages.

Using a fake data generator could be a solution, in case the dataset cannot be easily downloaded without setting up credentials:

https://github.com/pytorch/xla/blob/784b4d4f21751a54be0029a95f47d3896561c2a9/test/test_train_mp_mnist.py#L65

-->

Steps to reproduce the behavior:

1.
2.
3.

<!-- If you have a code sample, error messages, stack traces, please provide it here as well. Or better use the Colab template: https://github.com/pytorch/xla/blob/master/contrib/colab/issue-report.ipynb -->

## Expected behavior
func1 gives the graph:
```
HloModule SyncTensorsGraph.6, entry_computation_layout={(bf16[],bf16[1]{0})->(bf16[1]{0})}

ENTRY %SyncTensorsGraph.6 (p0: bf16[], p1: bf16[1]) -> (bf16[1]) {
  %p1 = bf16[1]{0} parameter(1), frontend_attributes={neff_input_name="input1"}, metadata={op_type="xla__device_data" op_name="xla__device_data"}
  %p0 = bf16[] parameter(0), frontend_attributes={neff_input_name="input0"}, metadata={op_type="xla__device_data" op_name="xla__device_data"}
  %broadcast = bf16[1]{0} broadcast(bf16[] %p0), dimensions={}, metadata={op_type="aten__mul" op_name="aten__mul"}
  %multiply = bf16[1]{0} multiply(bf16[1]{0} %p1, bf16[1]{0} %broadcast), metadata={op_type="aten__mul" op_name="aten__mul"}
  ROOT %tuple = (bf16[1]{0}) tuple(bf16[1]{0} %multiply), frontend_attributes={neff_output_names="output0"}
}

```

func2 gives a different graph:
```
HloModule SyncTensorsGraph.6, entry_computation_layout={(bf16[1]{0})->(bf16[1]{0})}

ENTRY %SyncTensorsGraph.6 (p0: bf16[1]) -> (bf16[1]) {
  %p0 = bf16[1]{0} parameter(0), frontend_attributes={neff_input_name="input0"}, metadata={op_type="xla__device_data" op_name="xla__device_data"}
  %constant = bf16[] constant(1), metadata={op_type="prim__Constant" op_name="prim__Constant"}
  %broadcast = bf16[1]{0} broadcast(bf16[] %constant), dimensions={}, metadata={op_type="aten__mul" op_name="aten__mul"}
  %multiply = bf16[1]{0} multiply(bf16[1]{0} %p0, bf16[1]{0} %broadcast), metadata={op_type="aten__mul" op_name="aten__mul"}
  ROOT %tuple = (bf16[1]{0}) tuple(bf16[1]{0} %multiply), frontend_attributes={neff_output_names="output0"}
}
```

func3 and func4 give the same graphs:
```
HloModule SyncTensorsGraph.14, entry_computation_layout={(bf16[],bf16[],bf16[1]{0})->(bf16[1]{0})}

ENTRY %SyncTensorsGraph.14 (p0: bf16[], p1: bf16[], p2: bf16[1]) -> (bf16[1]) {
  %p2 = bf16[1]{0} parameter(2), frontend_attributes={neff_input_name="input2"}, metadata={op_type="xla__device_data" op_name="xla__device_data"}
  %p1 = bf16[] parameter(1), frontend_attributes={neff_input_name="input1"}, metadata={op_type="xla__device_data" op_name="xla__device_data"}
  %broadcast.1 = bf16[1]{0} broadcast(bf16[] %p1), dimensions={}, metadata={op_type="aten__mul" op_name="aten__mul"}
  %multiply = bf16[1]{0} multiply(bf16[1]{0} %p2, bf16[1]{0} %broadcast.1), metadata={op_type="aten__mul" op_name="aten__mul"}
  %p0 = bf16[] parameter(0), frontend_attributes={neff_input_name="input0"}, metadata={op_type="xla__device_data" op_name="xla__device_data"}
  %broadcast.2 = bf16[1]{0} broadcast(bf16[] %p0), dimensions={}, metadata={op_type="aten__mul" op_name="aten__mul"}
  %multiply.1 = bf16[1]{0} multiply(bf16[1]{0} %multiply, bf16[1]{0} %broadcast.2), metadata={op_type="aten__mul" op_name="aten__mul"}
  %constant = bf16[] constant(1), metadata={op_type="prim__Constant" op_name="prim__Constant"}
  %reshape = bf16[1]{0} reshape(bf16[] %constant), metadata={op_type="aten__expand" op_name="aten__expand"}
  %broadcast = bf16[1]{0} broadcast(bf16[1]{0} %reshape), dimensions={0}, metadata={op_type="aten__expand" op_name="aten__expand"}
  %multiply.2 = bf16[1]{0} multiply(bf16[1]{0} %multiply.1, bf16[1]{0} %broadcast), metadata={op_type="aten__mul" op_name="aten__mul"}
  %subtract = bf16[1]{0} subtract(bf16[1]{0} %p2, bf16[1]{0} %multiply.2), metadata={op_type="aten__sub" op_name="aten__sub"}
  ROOT %tuple = (bf16[1]{0}) tuple(bf16[1]{0} %subtract), frontend_attributes={neff_output_names="output0"}
}
```

<!-- A clear and concise description of what you expected to happen. -->


## Environment

 - Reproducible on XLA backend [CPU/TPU]:
 - torch_xla version:


## Additional context

<!-- Add any other context about the problem here. -->

**Comment by JackCaoG** [comment]
Looking at the 4 functions, from pytorch/xla perspective, it is expected to generate different graph because the computation is different. 

For example in the first func it is 
```
    param.mul_(1 - lr * weight_decay)
```
while in last 2 graphs they are
```
    param.sub_(param * lr * weight_decay)
```

I guess the question is more of, what make AdamW to choose which branch to take? Is it some config passed down by user, or some runtime value.

**Comment by amithrm** [comment]
Correct, we had to change the first form into second form so that we get the same graphs. The first form (in first and second funcs) is yielding two different graphs based on the state of the optimizer. This is causing repeated compilations. 

To fix that, we had to change the optimizer (from third and fourth funcs). This code is giving the same graph outputs.

Can we fix the first form so that with any optimizer state, we see the same graph?

**Comment by JackCaoG** [comment]
Ok now I understand, you expect the func1 and func2 generate the same graph, but they are not.. Let me look into why that's the case.

**Comment by JackCaoG** [comment]
so if I update the test case to 
```
import torch
import torch_xla

import torch_xla.core.xla_model as xm



def func(lr):
    param = torch.FloatTensor([0.001]).to(xm.xla_device())
    weight_decay = 0.01
    param.mul_(1 - lr * weight_decay)
    #print(param)
    return param


res1 = func(2.9999999999999997e-06)
import pdb; pdb.set_trace()
print(torch_xla._XLAC._get_xla_tensors_text([res1]))
res2 = func(4.6874999999999995e-08)
print(torch_xla._XLAC._get_xla_tensors_text([res2]))
```

they yield the same ir
```
IR {
  %0 = f32[] xla::device_data(), xla_shape=f32[], device=CPU:0
  %1 = f32[1]{0} xla::device_data(), xla_shape=f32[1]{0}, device=CPU:0
  %2 = f32[1]{0} aten::mul(%1, %0), xla_shape=f32[1]{0}, ROOT=0
}

IR {
  %0 = f32[] prim::Constant(), xla_shape=f32[], value=1
  %1 = f32[1]{0} xla::device_data(), xla_shape=f32[1]{0}, device=CPU:0
  %2 = f32[1]{0} aten::mul(%1, %0), xla_shape=f32[1]{0}, ROOT=0
}
```

but if I changed to `print(torch_xla._XLAC._get_xla_tensors_hlo([res1]))`, they are actually different
```
ENTRY %IrToHlo.6 (p0.1: f32[], p1.2: f32[1]) -> (f32[1]) {
  %p1.2 = f32[1]{0} parameter(1)
  %p0.1 = f32[] parameter(0)
  %broadcast.3 = f32[1]{0} broadcast(f32[] %p0.1), dimensions={}
  %multiply.4 = f32[1]{0} multiply(f32[1]{0} %p1.2, f32[1]{0} %broadcast.3)
  ROOT %tuple.5 = (f32[1]{0}) tuple(f32[1]{0} %multiply.4)
}


HloModule IrToHlo.6, entry_computation_layout={(f32[1]{0})->(f32[1]{0})}

ENTRY %IrToHlo.6 (p0.2: f32[1]) -> (f32[1]) {
  %p0.2 = f32[1]{0} parameter(0)
  %constant.1 = f32[] constant(1)
  %broadcast.3 = f32[1]{0} broadcast(f32[] %constant.1), dimensions={}
  %multiply.4 = f32[1]{0} multiply(f32[1]{0} %p0.2, f32[1]{0} %broadcast.3)
  ROOT %tuple.5 = (f32[1]{0}) tuple(f32[1]{0} %multiply.4)
}
```

res1 takes an additional argument, where res2 just uses constant `1`. The reason is that on the value of `1 - lr * weight_decay`. If you print the value of them, they are `0.99999997` and `0.99999999953125`. The second one is too close to 1 so it got rounded to 1. Isn't there a flag we provided to disabled the special handling of 0/1? Let me find it.

**Comment by JackCaoG** [comment]
so if I run with `XLA_NO_SPECIAL_SCALARS=1` I can see the HLO being dumped is the same
```
HloModule IrToHlo.6, entry_computation_layout={(f32[],f32[1]{0})->(f32[1]{0})}

ENTRY %IrToHlo.6 (p0.1: f32[], p1.2: f32[1]) -> (f32[1]) {
  %p1.2 = f32[1]{0} parameter(1)
  %p0.1 = f32[] parameter(0)
  %broadcast.3 = f32[1]{0} broadcast(f32[] %p0.1), dimensions={}
  %multiply.4 = f32[1]{0} multiply(f32[1]{0} %p1.2, f32[1]{0} %broadcast.3)
  ROOT %tuple.5 = (f32[1]{0}) tuple(f32[1]{0} %multiply.4)
}
```

**Comment by amithrm** [comment]
Jack, yes I run with this flag on..strange, I see different graphs..

**Comment by amithrm** [comment]
The test case has this flag always on...

**Comment by JackCaoG** [comment]
https://github.com/pytorch/xla/blob/2855373cd016449f810989b0f9a6dd83b758cf8b/torch_xla/csrc/init_python_bindings.cpp#L750 is where we set the flag, and it needs to set before `torch_xla` import/init
